eth/handler.go

```diff
@@ -840,46 +838,39 @@ func (pm *ProtocolManager) handleFirehoseMsg(p *firehosePeer) error {
 
 	case GetStateNodesCode:
 		msgStream := rlp.NewStream(msg.Payload, uint64(msg.Size))
-		if _, err := msgStream.List(); err != nil {
-			return err
-		}
-		var reqID uint64
-		if err := msgStream.Decode(&reqID); err != nil {
-			return errResp(ErrDecode, "msg %v: %v", msg, err)
-		}
-
-		var blockHash common.Hash
-		if err := msgStream.Decode(&blockHash); err != nil {
+		var request getStateRangesOrNodes
+		if err := msgStream.Decode(&request); err != nil {
 			return errResp(ErrDecode, "msg %v: %v", msg, err)
 		}
 
-		if _, err := msgStream.List(); err != nil {
-			return err
-		}
+		n := len(request.Prefixes)
 
 		var response stateNodesMsg
-		response.ID = reqID
+		response.ID = request.ID
+		response.Nodes = make([][]byte, n)
 
-		block := pm.blockchain.GetBlockByHash(blockHash)
+		block := pm.blockchain.GetBlockByHash(request.Block)
```

Comment:
If you only need `block` to get the block number, there is a more efficient way of doing it, in `core/rawdb/accessor_chain.go`, func `ReadHeaderNumber`, however, it requires `DatabaseReader`. Perhaps it makes sense to access these things not via `pm.blockchain`, but instead have something like `pm.db`. This is for future improvements